ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision.hs

```diff
@@ -19,1133 +19,152 @@ module Ouroboros.Network.BlockFetch.Decision
+        -- current peer started (and a grace period), then the current peer is bad.
+        -- We push it at the end of the queue and demote it from CSJ dynamo.
+        & checkLastChainSelStarvation
+
+    -- Compute the actual block fetch decision. This contains only declines and
+    -- at most one request. 'theDecision' is therefore a 'Maybe'.
+    let (theDecision, declines) =
+          fetchDecisionsBulkSync
+            fetchDecisionPolicy
+            currentChain
+            fetchedBlocks
+            fetchedMaxSlotNo
+            peersOrder
+            candidatesAndPeers
+
+    -- If the peer that is supposed to fetch the block is not the current one in
+    -- the peers order, then we have shifted our focus: we make the new peer our
+    -- current one and we put back the previous current peer at the beginning of
+    -- the queue; not the end, because it has not done anything wrong.
+    case theDecision of
+      Just (_, (_, _, _, thePeer, _))
+        | Just thePeer /= peersOrderCurrent -> do
+            peersOrderStart' <- getMonotonicTime
+            let peersOrder' =
+                  PeersOrder
+                    { peersOrderCurrent = Just thePeer,
+                      peersOrderStart = peersOrderStart',
+                      peersOrderOthers = mcons peersOrderCurrent (filter (/= thePeer) peersOrderOthers)
+                    }
+            writePeersOrder peersOrder'
+      _ -> pure ()
+
+    pure $
+      maybe [] (singleton . first Right) theDecision
+        ++ map (first Left) declines
+    where
+      alignPeersOrderWithActualPeers :: PeersOrder peer -> [peer] -> PeersOrder peer
+      alignPeersOrderWithActualPeers
+        PeersOrder {peersOrderCurrent, peersOrderStart, peersOrderOthers}
+        actualPeers =
+          let peersOrderCurrent' = case peersOrderCurrent of
+                Just peersOrderCurrent_ | peersOrderCurrent_ `elem` actualPeers -> peersOrderCurrent
+                _ -> Nothing
+              peersOrderOthers' =
+                filter (`elem` actualPeers) peersOrderOthers
+                  ++ filter (`notElem` peersOrderOthers) actualPeers
+           in PeersOrder
+                { peersOrderCurrent = peersOrderCurrent',
+                  peersOrderOthers = peersOrderOthers',
+                  peersOrderStart
+                }
+
+      checkLastChainSelStarvation :: PeersOrder peer -> m (PeersOrder peer)
+      checkLastChainSelStarvation
+        peersOrder@PeersOrder {peersOrderCurrent, peersOrderStart, peersOrderOthers} =
+          case chainSelStarvation of
+            ChainSelStarvationEndedAt time
+              | time < addTime (1 :: DiffTime) peersOrderStart ->
+                pure peersOrder
+            _ -> do
+              let peersOrder' =
+                    PeersOrder
+                      { peersOrderCurrent = Nothing,
+                        peersOrderOthers = msnoc peersOrderOthers peersOrderCurrent,
+                        peersOrderStart
```

Comment:
The dynamo is being rotated but the start time is preserved. Unless this is updated elsewhere, I would expect:
```suggestion
                        time
```

Reply:
Oh, but it is updated by the caller `checkLastChainSelStarvation`. Please, disregard the above comment.